scio-core/src/main/scala-2.12/com/spotify/scio/coders/instances/TupleCoders.scala

```diff
@@ -0,0 +1,4715 @@
+          throw CoderStackTrace.append(
+            e,
+            Some(
+              s"Exception while trying to `${msg._1}` an instance" +
+                s" of Tuple3: Can't decode field ${msg._2}"
+            ),
+            materializationStackTrace
+          )
+      }
+
+    override def encode(value: (A, B, C), os: OutputStream): Unit = {
+      onErrorMsg("encode" -> "_1")(ac.encode(value._1, os))
+      onErrorMsg("encode" -> "_2")(bc.encode(value._2, os))
+      onErrorMsg("encode" -> "_3")(cc.encode(value._3, os))
+    }
+    override def decode(is: InputStream): (A, B, C) =
+      (
+        onErrorMsg("decode" -> "_1")(ac.decode(is)),
+        onErrorMsg("decode" -> "_2")(bc.decode(is)),
+        onErrorMsg("decode" -> "_3")(cc.decode(is))
+      )
+
+    override def toString: String =
+      s"Tuple3Coder(_1 -> $ac, _2 -> $bc, _3 -> $cc)"
+
+    // delegate methods for determinism and equality checks
+
+    override def verifyDeterministic(): Unit = {
+      val cs = List("_1" -> ac, "_2" -> bc, "_3" -> cc)
+      val problems = cs.flatMap { case (label, c) =>
+        try {
+          c.verifyDeterministic()
+          Nil
+        } catch {
+          case e: NonDeterministicException =>
+            val reason = s"field $label is using non-deterministic $c"
+            List(reason -> e)
+        }
+      }
+
+      problems match {
+        case (_, e) :: _ =>
+          val reasons = problems.map { case (reason, _) => reason }
+          throw new NonDeterministicException(this, reasons.asJava, e)
+        case Nil =>
+      }
+    }
+
+    override def consistentWithEquals(): Boolean =
+      ac.consistentWithEquals() && bc.consistentWithEquals() && cc.consistentWithEquals()
+
+    override def structuralValue(value: (A, B, C)): AnyRef =
+      if (consistentWithEquals()) {
+        value.asInstanceOf[AnyRef]
+      } else {
+        (ac.structuralValue(value._1), bc.structuralValue(value._2), cc.structuralValue(value._3))
+      }
+
+    // delegate methods for byte size estimation
+    override def isRegisterByteSizeObserverCheap(value: (A, B, C)): Boolean =
+      ac.isRegisterByteSizeObserverCheap(value._1) && bc.isRegisterByteSizeObserverCheap(
+        value._2
+      ) && cc.isRegisterByteSizeObserverCheap(value._3)
+
+    override def registerByteSizeObserver(
+      value: (A, B, C),
+      observer: ElementByteSizeObserver
+    ): Unit = {
+      ac.registerByteSizeObserver(value._1, observer)
+      bc.registerByteSizeObserver(value._2, observer)
+      cc.registerByteSizeObserver(value._3, observer)
+    }
+  }
+
+  implicit def tuple3Coder[A, B, C](implicit
+    CA: Strict[Coder[A]],
+    CB: Strict[Coder[B]],
+    CC: Strict[Coder[C]]
+  ): Coder[(A, B, C)] =
+    Coder.transform(CA.value) { ac =>
+      Coder.transform(CB.value) { bc =>
+        Coder.transform(CC.value)(cc => Coder.beam(new Tuple3Coder[A, B, C](ac, bc, cc)))
+      }
+    }
```

Comment:
Nesting `Coder.transform` will be slower than the code generated using magnolia which is why the old version was using `Coder.gen`.